In [24]:
import duckdb
import pandas as pd

In [25]:
con = duckdb.connect(r"C:\Users\admin\Desktop\YRES\Andi Dong - Summer Job Analysis\csj_analysis.duckdb")

print(con.execute("SHOW TABLES").df())

                    name
0  mart_ab_organizations
1        mart_ab_ridings
2  mart_bc_organizations
3        mart_bc_ridings
4  mart_on_organizations
5        mart_on_ridings
6                raw_csj
7                stg_csj


In [26]:
on_ridings = con.execute("SELECT * FROM mart_on_ridings").df()
on_orgs    = con.execute("SELECT * FROM mart_on_organizations").df()
bc_ridings = con.execute("SELECT * FROM mart_bc_ridings").df()
bc_orgs    = con.execute("SELECT * FROM mart_bc_organizations").df()
ab_ridings = con.execute("SELECT * FROM mart_ab_ridings").df()
ab_orgs    = con.execute("SELECT * FROM mart_ab_organizations").df()

In [27]:
print("ON ridings:", on_ridings.shape)
print("ON orgs:", on_orgs.shape)
print("BC ridings:", bc_ridings.shape)
print("BC orgs:", bc_orgs.shape)
print("AB ridings:", ab_ridings.shape)
print("AB orgs:", ab_orgs.shape)

ON ridings: (121, 29)
ON orgs: (3630, 29)
BC ridings: (42, 29)
BC orgs: (1260, 29)
AB ridings: (35, 29)
AB orgs: (1021, 29)


In [28]:
on_ridings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121 entries, 0 to 120
Data columns (total 29 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   constituency                   121 non-null    object 
 1   total_number_of_organizations  121 non-null    int64  
 2   funding_2017                   121 non-null    float64
 3   jobs_2017                      121 non-null    float64
 4   avg_salary_2017                121 non-null    float64
 5   funding_2018                   121 non-null    float64
 6   jobs_2018                      121 non-null    float64
 7   avg_salary_2018                121 non-null    float64
 8   funding_2019                   121 non-null    float64
 9   jobs_2019                      121 non-null    float64
 10  avg_salary_2019                121 non-null    float64
 11  funding_2020                   121 non-null    float64
 12  jobs_2020                      121 non-null    flo

In [29]:
on_ridings.head(100)

,constituency,total_number_of_organizations,funding_2017,jobs_2017,avg_salary_2017,funding_2018,jobs_2018,avg_salary_2018,funding_2019,jobs_2019,...,avg_salary_2022,funding_2023,jobs_2023,avg_salary_2023,funding_2024,jobs_2024,avg_salary_2024,funding_2025,jobs_2025,avg_salary_2025
0,Ajax,252,578737.0,192.0,3014.255208,572233.0,181.0,3161.508287,894419.0,218.0,...,5034.132597,1316304.0,330.0,3988.800000,1178030.0,289.0,4076.228374,1243957.0,276.0,4507.090580
1,Algoma - Manitoulin - Kapuskasing,532,577524.0,281.0,2055.245552,629683.0,247.0,2549.323887,680789.0,246.0,...,2789.140000,654138.0,208.0,3144.894231,589164.0,187.0,3150.609626,641060.0,191.0,3356.335079
2,Aurora - Oak Ridges - Richmond Hill,254,498342.0,182.0,2738.142857,551081.0,187.0,2946.957219,807266.0,259.0,...,3246.029817,918555.0,249.0,3688.975904,961185.0,238.0,4038.592437,1028222.0,252.0,4080.246032
3,Barrie - Innisfil,215,497071.0,190.0,2616.163158,557911.0,192.0,2905.786458,885464.0,309.0,...,2982.502381,989431.0,297.0,3331.417508,970066.0,265.0,3660.626415,1112313.0,263.0,4229.326996
4,Barrie - Springwater - Oro-Medonte,335,590532.0,194.0,3043.979381,620863.0,178.0,3487.994382,908163.0,287.0,...,2809.369388,1066989.0,317.0,3365.895899,1044756.0,290.0,3602.606897,1098419.0,292.0,3761.708904
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,Scarborough North,296,662934.0,239.0,2773.782427,769616.0,242.0,3180.231405,995343.0,274.0,...,4135.168317,954064.0,199.0,4794.291457,957129.0,188.0,5091.111702,1099161.0,213.0,5160.380282
96,Scarborough Southwest,250,655290.0,197.0,3326.345178,759747.0,223.0,3406.937220,1066265.0,275.0,...,4212.675000,1155983.0,260.0,4446.088462,1113707.0,231.0,4821.242424,1186854.0,258.0,4600.209302
97,Simcoe - Grey,413,507049.0,169.0,3000.289941,588131.0,222.0,2649.238739,797342.0,271.0,...,2633.709350,881535.0,283.0,3114.964664,907947.0,262.0,3465.446565,1100028.0,274.0,4014.700730
98,Simcoe North,364,563636.0,276.0,2042.159420,596407.0,202.0,2952.509901,866009.0,285.0,...,3178.149660,957278.0,287.0,3335.463415,970740.0,246.0,3946.097561,1136113.0,256.0,4437.941406


In [30]:
on_orgs.head(100)

,constituency,org_name,funding_2017,jobs_2017,avg_salary_2017,funding_2018,jobs_2018,avg_salary_2018,funding_2019,jobs_2019,...,avg_salary_2022,funding_2023,jobs_2023,avg_salary_2023,funding_2024,jobs_2024,avg_salary_2024,funding_2025,jobs_2025,avg_salary_2025
0,Essex,Town of Essex,16753.0,10.0,1675.300000,0.0,0.0,NaN,7350.0,5.0,...,1972.080000,20925.0,10.0,2092.500000,79012.0,35.0,2257.485714,96148.0,41.0,2345.073171
1,Essex,The Town of LaSalle,0.0,0.0,NaN,0.0,0.0,NaN,12096.0,6.0,...,1806.891892,44060.0,21.0,2098.095238,58184.0,26.0,2237.846154,0.0,0.0,NaN
2,Essex,Friends of Fort Malden,42562.0,12.0,3546.833333,44189.0,12.0,3682.416667,44560.0,17.0,...,4305.909091,32144.0,7.0,4592.000000,74188.0,15.0,4945.866667,78063.0,15.0,5204.200000
3,Essex,Municipality of Lakeshore,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,...,1989.961538,52146.0,27.0,1931.333333,66721.0,32.0,2085.031250,0.0,0.0,NaN
4,Essex,Kiwanis Club of Windsor,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,...,4397.045455,24446.0,5.0,4889.200000,99080.0,19.0,5214.736842,103364.0,19.0,5440.210526
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,Barrie - Innisfil,Connexus Church,31558.0,12.0,2629.833333,0.0,0.0,NaN,47578.0,12.0,...,4866.000000,60360.0,13.0,4643.076923,51250.0,10.0,5125.000000,59355.0,11.0,5395.909091
96,Barrie - Innisfil,Chabad Jewish Centre of Barrie,17710.0,6.0,2951.666667,14934.0,4.0,3733.500000,25815.0,7.0,...,4188.571429,23766.0,5.0,4753.200000,34739.0,7.0,4962.714286,52249.0,10.0,5224.900000
97,Barrie - Innisfil,Recreation and Culture Services,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,...,NaN,0.0,0.0,NaN,98793.0,43.0,2297.511628,0.0,0.0,NaN
98,Barrie - Innisfil,Innisfil Community Church,27762.0,8.0,3470.250000,0.0,0.0,NaN,54674.0,13.0,...,NaN,24355.0,5.0,4871.000000,50789.0,10.0,5078.900000,67401.0,13.0,5184.692308


In [31]:
# 第一个任务，把所有的avg salary全部转成保留两位小数形式！
avg_cols = [col for col in on_ridings.columns if 'avg_salary' in col]
on_ridings[avg_cols] = on_ridings[avg_cols].round(2)

avg_cols = [col for col in on_orgs.columns if 'avg_salary' in col]
on_orgs[avg_cols] = on_orgs[avg_cols].round(2)

avg_cols = [col for col in bc_ridings.columns if 'avg_salary' in col]
bc_ridings[avg_cols] = bc_ridings[avg_cols].round(2)

avg_cols = [col for col in bc_orgs.columns if 'avg_salary' in col]
bc_orgs[avg_cols] = bc_orgs[avg_cols].round(2)

avg_cols = [col for col in ab_ridings.columns if 'avg_salary' in col]
ab_ridings[avg_cols] = ab_ridings[avg_cols].round(2)

avg_cols = [col for col in ab_orgs.columns if 'avg_salary' in col]
ab_orgs[avg_cols] = ab_orgs[avg_cols].round(2)

In [32]:
on_ridings.head()

,constituency,total_number_of_organizations,funding_2017,jobs_2017,avg_salary_2017,funding_2018,jobs_2018,avg_salary_2018,funding_2019,jobs_2019,...,avg_salary_2022,funding_2023,jobs_2023,avg_salary_2023,funding_2024,jobs_2024,avg_salary_2024,funding_2025,jobs_2025,avg_salary_2025
0,Ajax,252,578737.0,192.0,3014.26,572233.0,181.0,3161.51,894419.0,218.0,...,5034.13,1316304.0,330.0,3988.80,1178030.0,289.0,4076.23,1243957.0,276.0,4507.09
1,Algoma - Manitoulin - Kapuskasing,532,577524.0,281.0,2055.25,629683.0,247.0,2549.32,680789.0,246.0,...,2789.14,654138.0,208.0,3144.89,589164.0,187.0,3150.61,641060.0,191.0,3356.34
2,Aurora - Oak Ridges - Richmond Hill,254,498342.0,182.0,2738.14,551081.0,187.0,2946.96,807266.0,259.0,...,3246.03,918555.0,249.0,3688.98,961185.0,238.0,4038.59,1028222.0,252.0,4080.25
3,Barrie - Innisfil,215,497071.0,190.0,2616.16,557911.0,192.0,2905.79,885464.0,309.0,...,2982.50,989431.0,297.0,3331.42,970066.0,265.0,3660.63,1112313.0,263.0,4229.33
4,Barrie - Springwater - Oro-Medonte,335,590532.0,194.0,3043.98,620863.0,178.0,3487.99,908163.0,287.0,...,2809.37,1066989.0,317.0,3365.90,1044756.0,290.0,3602.61,1098419.0,292.0,3761.71


In [33]:
stg_csj    = con.execute("SELECT * FROM stg_csj").df()
stg_csj.head() 

,year,region,constituency,org_name,amount,jobs
0,2025,Alberta,Banff - Airdrie,Airdrie & District Soccer Association,31621,7
1,2025,Alberta,Banff - Airdrie,Airdrie Chamber of Commerce,9341,2
2,2025,Alberta,Banff - Airdrie,Alberta Institute for Wildlife Conservation,74421,16
3,2025,Alberta,Banff - Airdrie,Ascend Pediatrics,1800,1
4,2025,Alberta,Banff - Airdrie,Banff Sport Medicine Foundation,4708,1


### STEP 1 先把用于Tableau可视化的数据集准备好！做好相关格式修改！ ### 

In [34]:
# 整体做清理，把省份column加上去！优先尝试把六个大表合并成两大类大表并带省份！
on_ridings['province'] = 'Ontario'
on_orgs['province'] = 'Ontario'
bc_ridings['province'] = 'British Columbia'
bc_orgs['province'] = 'British Columbia'
ab_ridings['province'] = 'Alberta'
ab_orgs['province'] = 'Alberta'

In [35]:
on_ridings.head(3)


,constituency,total_number_of_organizations,funding_2017,jobs_2017,avg_salary_2017,funding_2018,jobs_2018,avg_salary_2018,funding_2019,jobs_2019,...,funding_2023,jobs_2023,avg_salary_2023,funding_2024,jobs_2024,avg_salary_2024,funding_2025,jobs_2025,avg_salary_2025,province
0,Ajax,252,578737.0,192.0,3014.26,572233.0,181.0,3161.51,894419.0,218.0,...,1316304.0,330.0,3988.80,1178030.0,289.0,4076.23,1243957.0,276.0,4507.09,Ontario
1,Algoma - Manitoulin - Kapuskasing,532,577524.0,281.0,2055.25,629683.0,247.0,2549.32,680789.0,246.0,...,654138.0,208.0,3144.89,589164.0,187.0,3150.61,641060.0,191.0,3356.34,Ontario
2,Aurora - Oak Ridges - Richmond Hill,254,498342.0,182.0,2738.14,551081.0,187.0,2946.96,807266.0,259.0,...,918555.0,249.0,3688.98,961185.0,238.0,4038.59,1028222.0,252.0,4080.25,Ontario


In [36]:
# 所有ridings/orgs表格拼接到一起，是表示所有省份下所有信息！
all_ridings = pd.concat([on_ridings, bc_ridings, ab_ridings], ignore_index=True)
all_orgs = pd.concat([on_orgs, bc_orgs, ab_orgs], ignore_index=True)

In [37]:
print("all_ridings:", all_ridings.shape)
print("all_orgs:", all_orgs.shape)

all_ridings: (198, 30)
all_orgs: (5911, 30)


In [38]:
all_ridings_long = pd.melt(all_ridings, 
                           id_vars = ['constituency', 'province', 'total_number_of_organizations'],
                          var_name = 'metric_year',
                          value_name = 'value')

In [39]:
print(all_ridings_long.head(10))
print(all_ridings_long.shape)

                          constituency province  \
0                                 Ajax  Ontario   
1    Algoma - Manitoulin - Kapuskasing  Ontario   
2  Aurora - Oak Ridges - Richmond Hill  Ontario   
3                    Barrie - Innisfil  Ontario   
4   Barrie - Springwater - Oro-Medonte  Ontario   
5                        Bay of Quinte  Ontario   
6                  Beaches - East York  Ontario   
7                      Brampton Centre  Ontario   
8                        Brampton East  Ontario   
9                       Brampton North  Ontario   

   total_number_of_organizations   metric_year     value  
0                            252  funding_2017  578737.0  
1                            532  funding_2017  577524.0  
2                            254  funding_2017  498342.0  
3                            215  funding_2017  497071.0  
4                            335  funding_2017  590532.0  
5                            577  funding_2017  730243.0  
6                        

In [40]:
# 将metric列分开！
all_ridings_long[['metric','year']] = all_ridings_long['metric_year'].str.rsplit('_',n=1,expand=True) #注意！此处使用right split！因为avg_salary_XXXX，必须从右侧开始找dashline来分割！！

# 删掉原来的metric_year!
# all_ridings_long = all_ridings_long.drop(columns='metric_year')

#再pivot，让fundings/job/salary重新分为三列！
all_ridings_long = all_ridings_long.pivot_table(
    index = ['constituency', 'province', 'total_number_of_organizations', 'year'],
    columns = 'metric',
    values = 'value'
).reset_index()

print(all_ridings_long.head(5))
print(all_ridings_long.shape)

metric constituency          province  total_number_of_organizations  year  \
0        Abbotsford  British Columbia                            208  2017   
1        Abbotsford  British Columbia                            208  2018   
2        Abbotsford  British Columbia                            208  2019   
3        Abbotsford  British Columbia                            208  2020   
4        Abbotsford  British Columbia                            208  2021   

metric  avg_salary   funding   jobs  
0          3653.07  537001.0  147.0  
1          2690.23  382013.0  142.0  
2          4243.39  572858.0  135.0  
3          4154.58  531786.0  128.0  
4          3773.36  792406.0  210.0  
(1782, 7)


In [41]:
all_ridings_long

metric,constituency,province,total_number_of_organizations,year,avg_salary,funding,jobs
0,Abbotsford,British Columbia,208,2017,3653.07,537001.0,147.0
1,Abbotsford,British Columbia,208,2018,2690.23,382013.0,142.0
2,Abbotsford,British Columbia,208,2019,4243.39,572858.0,135.0
3,Abbotsford,British Columbia,208,2020,4154.58,531786.0,128.0
4,Abbotsford,British Columbia,208,2021,3773.36,792406.0,210.0
...,...,...,...,...,...,...,...
1777,York South - Weston,Ontario,243,2021,4503.46,1193418.0,265.0
1778,York South - Weston,Ontario,243,2022,4554.02,1229586.0,270.0
1779,York South - Weston,Ontario,243,2023,4474.43,899360.0,201.0
1780,York South - Weston,Ontario,243,2024,4866.81,880893.0,181.0


In [43]:
# 现在来做org表格！
all_orgs

,constituency,org_name,funding_2017,jobs_2017,avg_salary_2017,funding_2018,jobs_2018,avg_salary_2018,funding_2019,jobs_2019,...,funding_2023,jobs_2023,avg_salary_2023,funding_2024,jobs_2024,avg_salary_2024,funding_2025,jobs_2025,avg_salary_2025,province
0,Essex,Town of Essex,16753.0,10.0,1675.30,0.0,0.0,NaN,7350.0,5.0,...,20925.0,10.0,2092.50,79012.0,35.0,2257.49,96148.0,41.0,2345.07,Ontario
1,Essex,The Town of LaSalle,0.0,0.0,NaN,0.0,0.0,NaN,12096.0,6.0,...,44060.0,21.0,2098.10,58184.0,26.0,2237.85,0.0,0.0,NaN,Ontario
2,Essex,Friends of Fort Malden,42562.0,12.0,3546.83,44189.0,12.0,3682.42,44560.0,17.0,...,32144.0,7.0,4592.00,74188.0,15.0,4945.87,78063.0,15.0,5204.20,Ontario
3,Essex,Municipality of Lakeshore,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,...,52146.0,27.0,1931.33,66721.0,32.0,2085.03,0.0,0.0,NaN,Ontario
4,Essex,Kiwanis Club of Windsor,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,...,24446.0,5.0,4889.20,99080.0,19.0,5214.74,103364.0,19.0,5440.21,Ontario
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5906,Sherwood Park - Fort Saskatchewan,Trinity Baptist Church,9463.0,2.0,4731.50,0.0,0.0,NaN,21409.0,2.0,...,13772.0,2.0,6886.00,16309.0,2.0,8154.50,8673.0,1.0,8673.00,Alberta
5907,Sherwood Park - Fort Saskatchewan,Town of Whitecourt,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,...,0.0,0.0,NaN,0.0,0.0,NaN,30030.0,15.0,2002.00,Alberta
5908,Sherwood Park - Fort Saskatchewan,The Hastings Lake Lutheran Bible Camp Association,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,...,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,NaN,Alberta
5909,Sherwood Park - Fort Saskatchewan,Shiloh,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,...,0.0,0.0,NaN,0.0,0.0,NaN,0.0,0.0,NaN,Alberta


In [44]:
all_orgs_long = pd.melt(
    all_orgs,
    id_vars=['constituency', 'province', 'org_name'],
    var_name='metric_year',
    value_name='value'
)

all_orgs_long[['metric', 'year']] = all_orgs_long['metric_year'].str.rsplit('_', n=1, expand=True)

all_orgs_long = all_orgs_long.pivot_table(
    index=['constituency', 'province', 'org_name', 'year'],
    columns='metric',
    values='value'
).reset_index()

print(all_orgs_long.head(5))
print(all_orgs_long.shape)

metric constituency          province           org_name  year  avg_salary  \
0        Abbotsford  British Columbia  ABFM Thrift Store  2017         NaN   
1        Abbotsford  British Columbia  ABFM Thrift Store  2018         NaN   
2        Abbotsford  British Columbia  ABFM Thrift Store  2019         NaN   
3        Abbotsford  British Columbia  ABFM Thrift Store  2020         NaN   
4        Abbotsford  British Columbia  ABFM Thrift Store  2021         NaN   

metric  funding  jobs  
0           0.0   0.0  
1           0.0   0.0  
2           0.0   0.0  
3           0.0   0.0  
4           0.0   0.0  
(53199, 7)


In [45]:
all_orgs_long

metric,constituency,province,org_name,year,avg_salary,funding,jobs
0,Abbotsford,British Columbia,ABFM Thrift Store,2017,NaN,0.0,0.0
1,Abbotsford,British Columbia,ABFM Thrift Store,2018,NaN,0.0,0.0
2,Abbotsford,British Columbia,ABFM Thrift Store,2019,NaN,0.0,0.0
3,Abbotsford,British Columbia,ABFM Thrift Store,2020,NaN,0.0,0.0
4,Abbotsford,British Columbia,ABFM Thrift Store,2021,NaN,0.0,0.0
...,...,...,...,...,...,...,...
53194,York South - Weston,Ontario,Weston Village Senior Centre,2021,3990.0,11970.0,3.0
53195,York South - Weston,Ontario,Weston Village Senior Centre,2022,4200.0,12600.0,3.0
53196,York South - Weston,Ontario,Weston Village Senior Centre,2023,4340.0,13020.0,3.0
53197,York South - Weston,Ontario,Weston Village Senior Centre,2024,NaN,0.0,0.0


In [46]:
# 导出三个CSV来tableau做图可视化分析！
# 1. Tableau用的riding长表
all_ridings_long.to_csv(
    r'C:\Users\admin\Desktop\YRES\Andi Dong - Summer Job Analysis\tableau_ridings_long.csv',
    index=False
)

# 2. Tableau用的org长表（TOP30）
all_orgs_long.to_csv(
    r'C:\Users\admin\Desktop\YRES\Andi Dong - Summer Job Analysis\tableau_orgs_top30_long.csv',
    index=False
)

# 3. 全量数据（stg_csj，所有组织）
stg_csj.to_csv(
    r'C:\Users\admin\Desktop\YRES\Andi Dong - Summer Job Analysis\tableau_all_orgs_full.csv',
    index=False
)

print("三个CSV导出完成！")
print("1. tableau_ridings_long.csv")
print("2. tableau_orgs_top30_long.csv") 
print("3. tableau_all_orgs_full.csv")

三个CSV导出完成！
1. tableau_ridings_long.csv
2. tableau_orgs_top30_long.csv
3. tableau_all_orgs_full.csv


### 到此步骤，专门用于TABLEAU画图的三个CSV导出完成！全部做好！###

### 现在开始正式整SQL不擅长的地方，专门用PYTHON对SQL导出的数据集做精细化清洗和格式修正！ 属于python能做的真正价值！ ###

2. 导出专业Excel（6个sheet）
   → 自动命名sheet
   → 表头加粗
   → 冻结首行
   → 列宽自适应
3. 给amount列加货币格式
   → 578737 → $578,737
4. 给每张表加一个provincial summary行
   → 每个省的选区汇总底部加一行Total
6. 简单趋势预测
   → 用线性回归预测2026年funding
   → 加到Excel里作为额外sheet

In [47]:
pip install openpyxl

Note: you may need to restart the kernel to use updated packages.


In [48]:
pip install openpyxl scikit-learn

Note: you may need to restart the kernel to use updated packages.


In [53]:
from openpyxl import load_workbook
from openpyxl.styles import Font, PatternFill, Alignment
from openpyxl.utils import get_column_letter

output_path = r'C:\Users\admin\Desktop\YRES\Andi Dong - Summer Job Analysis\CSJ_Analysis_Output.xlsx'

# 定义6个sheet的配置
sheets = {
    'ON ridings':       on_ridings,
    'ON organizations': on_orgs,
    'BC ridings':       bc_ridings,
    'BC organizations': bc_orgs,
    'AB ridings':       ab_ridings,
    'AB organizations': ab_orgs
}

# Step 1: 用pandas先导出基础数据
with pd.ExcelWriter(output_path, engine='openpyxl') as writer:
    for sheet_name, df in sheets.items():
        df.to_excel(writer, sheet_name=sheet_name, index=False)

# Step 2: 用openpyxl加格式
wb = load_workbook(output_path)

for sheet_name in wb.sheetnames:
    ws = wb[sheet_name]
    
    # 表头加粗 + 背景色 + 字体颜色
    header_font = Font(bold=True, color='FFFFFF', name='Arial')
    header_fill = PatternFill('solid', start_color='2F5597')  # 深蓝色
    
    for cell in ws[1]:
        cell.font = header_font
        cell.fill = header_fill
        cell.alignment = Alignment(horizontal='center', wrap_text=True)
    
    # 冻结首行
    ws.freeze_panes = 'A2'
    
    # 列宽自适应
    for col in ws.columns:
        max_length = 0
        col_letter = get_column_letter(col[0].column)
        for cell in col:
            if cell.value:
                max_length = max(max_length, len(str(cell.value)))
        ws.column_dimensions[col_letter].width = min(max_length + 2, 30)
    
    # 货币格式：所有funding列
    for row in ws.iter_rows(min_row=2):
        for cell in row:
            col_name = ws.cell(1, cell.column).value
            if col_name and 'funding' in str(col_name):
                cell.number_format = '$#,##0'
            elif col_name and 'avg_salary' in str(col_name):
                cell.number_format = '$#,##0.00'



wb.save(output_path)
print("专业Excel导出完成！")

'''
**代码做了什么：**
- 深蓝色表头 + 白色加粗字体
- 冻结首行，滚动时表头不动
- 列宽自适应，最宽30
- funding列加`$`货币格式
- avg_salary列加`$`保留两位小数
'''

专业Excel导出完成！


'\n**代码做了什么：**\n- 深蓝色表头 + 白色加粗字体\n- 冻结首行，滚动时表头不动\n- 列宽自适应，最宽30\n- funding列加`$`货币格式\n- avg_salary列加`$`保留两位小数\n'

### 最后来一步骤，简单做一个线性回归预测模型！目的在于预测所有的选区，记得不是机构！总共190多个选区下，根据17-25年所有数据，来分析每个选区的未来funding走势来判断后续可以理论性重点考虑投资哪一个选区最为合适？

In [54]:
stg_csj

,year,region,constituency,org_name,amount,jobs
0,2025,Alberta,Banff - Airdrie,Airdrie & District Soccer Association,31621,7
1,2025,Alberta,Banff - Airdrie,Airdrie Chamber of Commerce,9341,2
2,2025,Alberta,Banff - Airdrie,Alberta Institute for Wildlife Conservation,74421,16
3,2025,Alberta,Banff - Airdrie,Ascend Pediatrics,1800,1
4,2025,Alberta,Banff - Airdrie,Banff Sport Medicine Foundation,4708,1
...,...,...,...,...,...,...
137516,2017,Alberta,Yellowhead,Yellow Submarine,6043,3
137517,2017,Alberta,Yellowhead,Yellowhead County Library Board,7394,2
137518,2017,Alberta,Yellowhead,Yellowhead Gas Co-Op Ltd,6510,1
137519,2017,Alberta,Yellowhead,fRI Research,4392,1


In [55]:
# 按照省份，选区，年来分组！
constituency_yearly = stg_csj.groupby(['region','constituency','year']).agg(total_funding = ('amount','sum'), total_jobs = ('jobs', 'sum')).reset_index()

In [56]:
constituency_yearly

,region,constituency,year,total_funding,total_jobs
0,Alberta,Banff - Airdrie,2017,369399,117
1,Alberta,Banff - Airdrie,2018,371332,132
2,Alberta,Banff - Airdrie,2019,731178,215
3,Alberta,Banff - Airdrie,2020,1038083,234
4,Alberta,Banff - Airdrie,2021,1590778,467
...,...,...,...,...,...
1769,Ontario,York South - Weston,2021,1193418,265
1770,Ontario,York South - Weston,2022,1229586,270
1771,Ontario,York South - Weston,2023,899360,201
1772,Ontario,York South - Weston,2024,880893,181


In [57]:
constituency_yearly.head(50)

,region,constituency,year,total_funding,total_jobs
0,Alberta,Banff - Airdrie,2017,369399,117
1,Alberta,Banff - Airdrie,2018,371332,132
2,Alberta,Banff - Airdrie,2019,731178,215
3,Alberta,Banff - Airdrie,2020,1038083,234
4,Alberta,Banff - Airdrie,2021,1590778,467
5,Alberta,Banff - Airdrie,2022,1281065,375
6,Alberta,Banff - Airdrie,2023,748275,211
7,Alberta,Banff - Airdrie,2024,913255,271
8,Alberta,Banff - Airdrie,2025,613802,201
9,Alberta,Battle River - Crowfoot,2017,860600,263


In [58]:
from sklearn.linear_model import LinearRegression
import numpy as np

results = []

#每个选区group单独训练模型！
for (region, constituency), group in constituency_yearly.groupby(['region', 'constituency']):
    X = group['year'].values.reshape(-1,1) # year as features!
    y_funding = group['total_funding'].values # funding as the target 1
    y_jobs = group['total_jobs'].values # jobs created as the target 2

    # train the model
    model_funding = LinearRegression().fit(X, y_funding)
    model_jobs = LinearRegression().fit(X, y_jobs)

    # predict for values in 2026
    pred_funding = model_funding.predict([[2026]])[0]
    pred_jobs = model_jobs.predict([[2026]])[0]

    results.append({
    'region': region,
    'constituency': constituency,
    'pred_funding_2026': round(max(pred_funding, 0)),
    'pred_jobs_2026': round(max(pred_jobs, 0))
    })

predictions = pd.DataFrame(results)
print(predictions.head(10))

    region             constituency  pred_funding_2026  pred_jobs_2026
0  Alberta          Banff - Airdrie            1090843             321
1  Alberta  Battle River - Crowfoot            1114701             329
2  Alberta                Bow River             908803             260
3  Alberta           Calgary Centre            1590974             416
4  Alberta    Calgary Confederation            2167547             622
5  Alberta      Calgary Forest Lawn            1235810             333
6  Alberta         Calgary Heritage            1342974             321
7  Alberta        Calgary Midnapore            1446217             404
8  Alberta        Calgary Nose Hill             940869             223
9  Alberta      Calgary Rocky Ridge            1266578             342


In [60]:
predictions.shape

(198, 4)

In [61]:
with pd.ExcelWriter(output_path, mode='a', engine='openpyxl') as writer:
    predictions.to_excel(writer, sheet_name='2026 Predictions', index=False)

print("预测sheet添加完成！")

预测sheet添加完成！


In [62]:
all_ridings_long

metric,constituency,province,total_number_of_organizations,year,avg_salary,funding,jobs
0,Abbotsford,British Columbia,208,2017,3653.07,537001.0,147.0
1,Abbotsford,British Columbia,208,2018,2690.23,382013.0,142.0
2,Abbotsford,British Columbia,208,2019,4243.39,572858.0,135.0
3,Abbotsford,British Columbia,208,2020,4154.58,531786.0,128.0
4,Abbotsford,British Columbia,208,2021,3773.36,792406.0,210.0
...,...,...,...,...,...,...,...
1777,York South - Weston,Ontario,243,2021,4503.46,1193418.0,265.0
1778,York South - Weston,Ontario,243,2022,4554.02,1229586.0,270.0
1779,York South - Weston,Ontario,243,2023,4474.43,899360.0,201.0
1780,York South - Weston,Ontario,243,2024,4866.81,880893.0,181.0


In [64]:
# 把predictions整理成和all_ridings_long一样的行格式
pred_rows = predictions.copy()
pred_rows = pred_rows.rename(columns={
    'region': 'province',
    'pred_funding_2026': 'funding',
    'pred_jobs_2026': 'jobs'
})
pred_rows['year'] = 2026
pred_rows['avg_salary'] = (pred_rows['funding'] / pred_rows['jobs']).round(2)
pred_rows['total_number_of_organizations'] = None
pred_rows['is_predicted'] = True

# 原始数据加标记
all_ridings_long['is_predicted'] = False

# 拼接
all_ridings_long = pd.concat(
    [all_ridings_long, pred_rows[all_ridings_long.columns]], 
    ignore_index=True
)

print(all_ridings_long.shape)  # 应该是1980行

(1980, 8)


In [65]:
all_ridings_long

,constituency,province,total_number_of_organizations,year,avg_salary,funding,jobs,is_predicted
0,Abbotsford,British Columbia,208,2017,3653.07,537001.0,147.0,False
1,Abbotsford,British Columbia,208,2018,2690.23,382013.0,142.0,False
2,Abbotsford,British Columbia,208,2019,4243.39,572858.0,135.0,False
3,Abbotsford,British Columbia,208,2020,4154.58,531786.0,128.0,False
4,Abbotsford,British Columbia,208,2021,3773.36,792406.0,210.0,False
...,...,...,...,...,...,...,...,...
1975,Windsor - Tecumseh,Ontario,None,2026,4231.10,1421651.0,336.0,True
1976,Windsor West,Ontario,None,2026,4772.74,1765914.0,370.0,True
1977,York - Simcoe,Ontario,None,2026,4719.38,1156248.0,245.0,True
1978,York Centre,Ontario,None,2026,5326.29,1054606.0,198.0,True


In [66]:
all_ridings_long.to_csv(
    r'C:\Users\admin\Desktop\YRES\Andi Dong - Summer Job Analysis\tableau_ridings_long.csv',
    index=False
)
print("导出完成！")

导出完成！


# Python Layer - CSJ Analysis Pipeline
**目标：SQL做不到的自动化处理，为Tableau和YRES交付做准备**

---

## Step 1: 连接DuckDB，读取6张mart表
- 连接 `csj_analysis.duckdb`
- 读取 `mart_on/bc/ab_ridings` 和 `mart_on/bc/ab_organizations`
- 验证数据形状：ON(121,29) BC(42,29) AB(35,29)

---

## Step 2: 数据清洗
- 找出所有 `avg_salary` 列
- 统一四舍五入到2位小数
- 用for loop批量处理6张表

---

## Step 3: 加province列 + concat拼接
- 6张表各自加 `province` 列标记省份
- riding三表合并 → `all_ridings` (198行 × 30列)
- org三表合并   → `all_orgs`   (5911行 × 30列)

---

## Step 4: 宽表转长表（pd.melt + pivot_table）
**为什么要转：Tableau需要year作为独立列才能画趋势图**

- `pd.melt()` 把年份列转成行
- `str.rsplit('_', n=1)` 拆分metric和year
- `pivot_table()` 把funding/jobs/avg_salary变回三列
- 结果：
  - `all_ridings_long` (1782行 × 7列)
  - `all_orgs_long`    (53199行 × 7列)

---

## Step 5: 线性回归预测2026
**为什么用线性回归：每个选区只有9年数据，复杂模型会过拟合**

- 从 `stg_csj` 用pandas groupby得到选区年度汇总
- 对198个选区各自训练独立的LinearRegression模型
- X = year, y = total_funding / total_jobs
- 预测2026年每个选区的funding和jobs
- `max(pred, 0)` 确保预测值不为负

---

## Step 6: 预测结果整合
- 预测结果加入Excel作为第7个sheet（`2026 Predictions`）
- 预测结果转成长表格式，加 `is_predicted=True` 标记
- 拼接进 `all_ridings_long`
- 最终：1980行（1782历史 + 198预测）

---

## Step 7: 导出Excel（YRES交付）
- 6个sheet对应6张分析表
- 格式：深蓝表头 + 白色加粗字体 + 冻结首行 + 列宽自适应
- funding列：`$#,##0` 货币格式
- avg_salary列：`$#,##0.00` 格式
- 第7个sheet：2026预测结果

---

## Step 8: 导出Tableau CSV（3个文件）
| 文件名 | 内容 | 用途 |
|--------|------|------|
| `tableau_ridings_long.csv` | 198选区 × 10年(含2026预测) | 趋势图/预测对比 |
| `tableau_orgs_top30_long.csv` | TOP30组织历史数据 | 竞争对手分析 |
| `tableau_all_orgs_full.csv` | 全量原始数据 | 深度分析 |

---

## 技术栈总结
```
duckdb      → 连接数据库
pandas      → 数据处理（groupby/melt/pivot/concat/merge）
openpyxl    → Excel格式美化
sklearn     → 线性回归预测
```